### DISCUSSION TITLE
Auto Complete How to Store Free Text & How does Property Name Work?

### DISCUSSION THREAD

**Discussion by chobo2** [vikramlearning/blazorbootstrap#442] [Q&A]
Hi

I am looking at auto complete and have some questions.

My first question is Say I have this object

```
public class Customer2{
  public int Id {get; set;}
  public string CustomerName {get; set;}
}

<div class="row">
    <div class="col-md-5 col-sm-12">
        <AutoComplete @bind-Value="customerName"
                      TItem="Customer2"
                      DataProvider="CustomersDataProvider"
                      PropertyName="CustomerName"
                      Placeholder="Search a customer..."
                      OnChanged="(Customer2 customer) => OnAutoCompleteChanged(customer)" />
    </div>
</div>

@code {
    private string customerName;

    [Inject] ICustomerService _customerService { get; set; }

    private async Task<AutoCompleteDataProviderResult<Customer2>> CustomersDataProvider(AutoCompleteDataProviderRequest<Customer2> request)
    {
        var customers = await _customerService.GetCustomersAsync(request.Filter, request.CancellationToken); // API call
        return await Task.FromResult(new AutoCompleteDataProviderResult<Customer2> { Data = customers, TotalCount = customers.Count() });
    }

    private void OnAutoCompleteChanged(Customer2 customer)
    {
        // TODO: handle your own logic

        // NOTE: do null check
        Console.WriteLine($"'{customer?.CustomerName}' selected.");
    }
}
```

What happens if I want to capture the "id" instead of the "Name" do I have to do something like this

```
private Customer2 SelectedCustomer {get; set;}

    private void OnAutoCompleteChanged(Customer2 customer)
    {
       SelectedCustomer = customer;
        // TODO: handle your own logic

        // NOTE: do null check
        Console.WriteLine($"'{customer?.CustomerName}' selected.");
    }
```

My next question is what happens if I want to capture a "non found" result. Say I have these names

- John Smith
- Jim Brown
- Mike G

Yet, the type in "Sara Key". No results would be found, but I still would like to capture that input. HOw would I do this. I would have hoped it would be in the "customerName" variable but it does not seem to be the case.

Thanks

**Comment by gvreddy04** [comment]
> Hi
> 
> 
> 
> I am looking at auto complete and have some questions.
> 
> 
> 
> My first question is Say I have this object
> 
> 
> 
> ```
> 
> public class Customer2{
> 
>   public int Id {get; set;}
> 
>   public string CustomerName {get; set;}
> 
> }
> 
> 
> 
> <div class="row">
> 
>     <div class="col-md-5 col-sm-12">
> 
>         <AutoComplete @bind-Value="customerName"
> 
>                       TItem="Customer2"
> 
>                       DataProvider="CustomersDataProvider"
> 
>                       PropertyName="CustomerName"
> 
>                       Placeholder="Search a customer..."
> 
>                       OnChanged="(Customer2 customer) => OnAutoCompleteChanged(customer)" />
> 
>     </div>
> 
> </div>
> 
> 
> 
> @code {
> 
>     private string customerName;
> 
> 
> 
>     [Inject] ICustomerService _customerService { get; set; }
> 
> 
> 
>     private async Task<AutoCompleteDataProviderResult<Customer2>> CustomersDataProvider(AutoCompleteDataProviderRequest<Customer2> request)
> 
>     {
> 
>         var customers = await _customerService.GetCustomersAsync(request.Filter, request.CancellationToken); // API call
> 
>         return await Task.FromResult(new AutoCompleteDataProviderResult<Customer2> { Data = customers, TotalCount = customers.Count() });
> 
>     }
> 
> 
> 
>     private void OnAutoCompleteChanged(Customer2 customer)
> 
>     {
> 
>         // TODO: handle your own logic
> 
> 
> 
>         // NOTE: do null check
> 
>         Console.WriteLine($"'{customer?.CustomerName}' selected.");
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> What happens if I want to capture the "id" instead of the "Name" do I have to do something like this
> 
> 
> 
> ```
> 
> private Customer2 SelectedCustomer {get; set;}
> 
> 
> 
>     private void OnAutoCompleteChanged(Customer2 customer)
> 
>     {
> 
>        SelectedCustomer = customer;
> 
>         // TODO: handle your own logic
> 
> 
> 
>         // NOTE: do null check
> 
>         Console.WriteLine($"'{customer?.CustomerName}' selected.");
> 
>     }
> 
> ```

This is correct.

> 
> My next question is what happens if I want to capture a "non found" result. Say I have these names
> 
> 
> 
> - John Smith
> 
> - Jim Brown
> 
> - Mike G
> 
> 
> 
> Yet, the type in "Sara Key". No results would be found, but I still would like to capture that input. HOw would I do this. I would have hoped it would be in the "customerName" variable but it does not seem to be the case.
> 
> 
> 
> Thanks
> 
> 

This is not supported. May be you can show a button beside the auto complete to add the missing one using a modal or offcanvas.
